src/manifolds/SymplecticStiefel.jl

```diff
@@ -0,0 +1,554 @@
+
+TODO: Complete.
+Given the Riemannian metric at ``p``, ``g_p \colon `` we compute a mapping
+``c : T_p\operatorname{SpSt}(2n, 2k) \rightarrow ℝ^{2n \times 2k}``
+Calculated myself, but tailored to the right-invariant inner product of Bendokat-Zimmermann.
+Does not have the correct range to be a 'change_representer!'-mapping.
+"""
+function make_metric_compatible!(
+    ::SymplecticStiefel{n,k},
+    Y,
+    ::EuclideanMetric,
+    p,
+    X,
+) where {n,k}
+    Q = SymplecticMatrix(p, X)
+
+    pT_p = p' * p
+    pT_Q = p' * Q
+    inner_term = pT_Q' * (lu(pT_p) \ pT_Q)  # ∈ ℝ^{2n × 2n}
+
+    divisor = lu!(Array(add_scaled_I!(lmul!(-0.5, inner_term), 1.0)'))
+    mul!(Y, X, pT_p)
+    ldiv!(divisor, Y)
+    return Y
+end
+
+@doc raw"""
+    rand(M::SymplecticStiefel{n, k})
+
+Generate a random point ``p \in \operatorname{SpSt}(2n, 2k)``
+by first generating a random symplectic matrix
+``p_{\operatorname{Sp}} \in \operatorname{Sp}(2n)``,
+and then projecting onto the Symplectic Stiefel manifold using the
+[`canonical_projection`](@ref).
+That is, ``p = π(p_{\operatorname{Sp}})``
+"""
+function Base.rand(M::SymplecticStiefel{n,k}, hamiltonian_norm=1 / 2) where {n,k}
+    p_symplectic = rand(Symplectic(2n), hamiltonian_norm)
+    return canonical_projection(M, p_symplectic)
+end
+
+@doc raw"""
+    rand(::SymplecticStiefel{n, k}, p)
+
+The symplectic Stiefel tangent space at ``p`` can be parametrized as [^Bendokat2021]
+````math
+    T_p\operatorname{SpSt}(2n, 2k) = \{X = pΩ + p^sB \;|\;
+        Ω ∈ ℝ^{2k × 2k}, Ω^+ = -Ω,
+        p^s ∈ \operatorname{SpSt}(2n, 2(n- k)), B ∈ ℝ^{2(n-k) × 2k}, \},
+````
+where ``Ω \in `` is Hamiltonian and ``p^s`` means the symplectic complement of ``p`` s.t.
+``p^{+}p^{s} = 0``.
+
+To then generate random tangent vectors at ``p``, we set ``B = 0`` and generate a random
+Hamiltonian matrix ``Ω``.
+"""
+function Base.rand(::SymplecticStiefel{n,k}, p::AbstractMatrix) where {n,k}
+    Ω = rand_hamiltonian(Symplectic(2k))
+    return p * Ω
+end
+
+@doc raw"""
+    canonical_projection(::SymplecticStiefel, p_Sp)
+    canonical_projection!(::SymplecticStiefel{n,k}, p, p_Sp)
+
+Define the canonical projection from ``\operatorname{Sp}(2n, 2n)`` onto
+``\operatorname{SpSt}(2n, 2k)``, by projecting onto the first ``k`` columns
+and the ``n + 1``'th onto the ``n + k``'th columns [^Bendokat2021].
+
+It is assumed that the point ``p`` is on ``\operatorname{Sp}(2n, 2n)``.
+"""
+function canonical_projection(M::SymplecticStiefel{n,k}, p_Sp) where {n,k}
+    p_SpSt = similar(p_Sp, (2n, 2k))
+    return canonical_projection!(M, p_SpSt, p_Sp)
+end
+
+function canonical_projection!(::SymplecticStiefel{n,k}, p, p_Sp) where {n,k}
+    p[:, (1:k)] .= p_Sp[:, (1:k)]
+    p[:, ((k + 1):(2k))] .= p_Sp[:, ((n + 1):(n + k))]
+    return p
+end
+
+# compute p^+q (which is 2kx2k) in place of A
+function symplectic_inverse_times(M::SymplecticStiefel{n,k}, p, q) where {n,k}
```

Comment:
This should have a documentation (and be exported maybe)

Reply:
I've drafted documentation for this function now, and also exported `symplectic_inverse_times(::SymplecticStiefel, p, q)`
and `symplectic_inverse_times!(::SymplecticStiefel, A, p, q)`.